distributed/shuffle/_shuffle.py

```diff
@@ -103,9 +105,15 @@ def rearrange_by_column_p2p(
     column: str,
     npartitions: int | None = None,
 ) -> DataFrame:
+    import pandas as pd
+
     from dask.dataframe.core import new_dd_object
 
     meta = df._meta
+    if not pd.api.types.is_integer_dtype(meta[column]):
+        raise TypeError(
+            f"Expected meta {column=} to be an integer column, is {meta[column].dtype}."
+        )
```

Comment:
This will raise early if the meta column is not an integer. The shuffle internals rely on correct typing and the errors that are caused by type mismatches are pretty tricky (i.e., the shuffle will run but the output will be empty).